[ENHANCEMENT] Implement Exec Migration / Script Runner

Glenn-Ewing opened the pull request:
**Primary Change Description**
- Implement ExecMigration to allow for post a migrations "Script Runner" for data seeds, updates, etc.
- ExecMigrations are transaction based, so they will rollback unless all of the migration succeeds
- Adds/Updates all respective tests to include the new migration type
- Adds the concept of non-reversible and non-reducible Operations

**Related/Interdependent Changes**
- Refactors Operations to use separate Interfaces to indicate their functional capabilities
  (e.g. Reversible Operations and/or Reducible Operations)
- Refactors Migration/Operation Processors to indicate handling of all operation types while providing conditional processing based on the functional capabilities of the operations reported by the operations themselves

**Unrelated but Important Changes**
- Renames the class "Migration" to "TableMigration" (for better clarity of purpose)
- Adds a "Migration" alias class which extends TableMigration for backwards compatibility.
   Note: This will be deprecated in a future release.
- Updates composer to reflect min PHP standards of 7.4 or above
- Updates/unifies syntax standards for interfaces, parameters, type hinting, etc. to PHP 7.4 standards

- All tests pass

Note: Due to the nature of seed & update scripts, these are non reversible migrations, which is why the added transactions based support was necessary to ensure a partial script run would not persist, only a fully successful run.

Glenn-Ewing committed:
Implement ExecMigration

Implement ExecMigration to allow for transactional post migration Data Seeds, Migration, Updates, etc.

Glenn-Ewing committed:
Update MysqlHandlerTest.php

Glenn-Ewing committed:
Test Updates

Glenn-Ewing committed:
Syntax Update

Glenn-Ewing reviewed (commented):
Self-review completed.

Glenn-Ewing commented:
@joshmcrae Do we want to enforce a minimum PHP version compatibility for this library? I am ultimately just wondering if we should update the syntax to 7.4 standards, etc. If so, let me know; I'll be happy to do so.

joshmcrae commented on src/Handler.php:
```diff
@@ -104,18 +105,45 @@ private function processOperations(array $operations, array $versions, bool $red
             $sql = $this->getBuilder()->build($operation);
+
+            // Determine if transaction is required.
+            $transactionRequired = $operations instanceof ExecOperation;
```
This line should be performing `instanceof` against `$operation` instead of the array.

Glenn-Ewing replied:
@joshmcrae lol, yes... yes it should... thanks for that.

Glenn-Ewing replied:
Resolved!

joshmcrae commented on src/Handler.php:
```diff
@@ -104,18 +105,45 @@ private function processOperations(array $operations, array $versions, bool $red
+
+            // Begin trx if needed.
+            if ($transactionRequired) {
+                var_dump('YO!');
```
Try not to mix terminology in comments. Single line comments don't need a period (check previous comment in same block too).

There's a remnant debugging log here too.

Glenn-Ewing replied:
Oops! Thanks, Handled!

joshmcrae commented on src/Statement/MysqlStatementBuilder.php:
```diff
@@ -271,6 +274,26 @@ public function buildFunction(FunctionOperation $operation): string
+    {
+        switch ($operation->getOperation()) {
+            case ExecOperation::EXEC:
+                return sprintf(
```
The `sprintf` and use of a constant is redundant here. The result of `$operation-getBody()` may be returned directly instead.

Glenn-Ewing replied:
Gotcha, handled!

joshmcrae commented on src/Operation/ExecOperation.php:
```diff
@@ -0,0 +1,87 @@
+            throw new InvalidArgumentException('Cannot apply operations for a different execution.');
+        }
+
+        if ($this->getOperation() !== self::EXEC) {
```
I think the entire use of this class constant could be avoided since it's the only value used anywhere to represent an `ExecOperation`'s operation.

To that end I would remove the constant along with the `$operation` member variable (and it's constructor argument) so that `ExecOperation::getOperation()` always returns the static string `execute`.

This line would thus be `if ($operation->getOperation() !== $this->getOperation()) {`. At the moment it's checking it's own value and always returns true.

Glenn-Ewing replied:
With the only operation available being execute, I removed the argument as a whole from both the migration and operation classes, and updated the getOperation() method to return a static string as suggested. Let me know if you agree with the approach or not.

joshmcrae commented on src/Operation/ExecOperation.php:
```diff
@@ -0,0 +1,87 @@
+        return new ExecOperation(
+            $this->getName(),
+            $this->getOperation(),
+            $operation->getBody()
```
Would this simply overwrite the current operation?

Glenn-Ewing replied:
@joshmcrae This, in the other classes, allows for additions, udpates, etc. but yes in this case, I wonder if its a necessary method. The intent I had here would be that the body would just be overwritten in a merge, but perhaps we should have a script[] and push any updates to it? What do you think?

Glenn-Ewing replied:
@joshmcrae I think for now, simply overwriting this may be the simplest solution, I am not sure I would event allow for an apply method if it wasn't a required method. I considered comparing the body and throwing an exception if they did not match so that the exec names would need to be unique. What do you think of that idea?

joshmcrae replied:
I agree that `apply()` shouldn't even exist for this kind of operation. That said, it isn't technically a requirement yet because it's not enforced by `AbstractOperation`, it's just that everywhere an operation class is used it's assumed that it has that method.

If you have a look at `History::reduce()` it's type-hinted to accept `TableOperation[]` but what's really passed into it is a bunch of different operation types. `History` assumes the use of `TableOperation` in a lot of places and hasn't been refactored to deal more generally with `AbstractOperation`.

That's an issue that needs addressing so if we can do it as part of this PR that'd be great.

I'm thinking each concrete operation class will have to be able to signal whether is supports `apply()` or not, so that we can have these kind of operations that don't compose well.

Glenn-Ewing replied:
@joshmcrae understood, let me know if my refactor hits the mark for these concerns. 
FYI All parameters will no longer hint to AbstractOperation rather, they will hint to OperationInterface as this seemed more intuitive.

joshmcrae reviewed (changes requested):
Hi @Glenn-Ewing!

Thank you for this much needed enhancement. I've left a few comments below for your review.

Glenn-Ewing commented:
@joshmcrae Apologies, some of this may me on autopilot trying to make things work a bit too hastily. I'll swing back again and get these buttoned up.

joshmcrae commented:
@Glenn-Ewing

Regarding minimum PHP version, please check that all of our projects using the library are on at least 7.4. If so, then I'm happy for more modern language features to be used and the version to be bumped. It will require a major release.

Glenn-Ewing committed:
Since execute is the only supported option, it seems logical to just remove the operation type parameter.

Glenn-Ewing committed:
Comment cleanup

Glenn-Ewing committed:
Fix comparator

Glenn-Ewing committed:
Operation Refactor and Syntax Updates

- Refactors Operations to use separate Interfaces to indicate their functional capabilities.
- Refactors Migration/Operation Processors to indicate handling of all operation types with conditional processing based on the fnuctional capabilities of the operations reported by the operations themselves
- Renames the Migration class to TableMigration (for better calrity of purpose)
- Adds a Migration alais class which extends TableMigration for backwards compatibility. Note: This will be deprecated in a future release.
- Updates composer to reflect min PHP standards of 7.4 or above
- Updates/unifies syntax standards for interfaces, parameters, type hinting, etc. to PHP 7.4 standards

Glenn-Ewing commented:
@joshmcrae Please review when you get a moment. Per our discussion, changes have been made as follows:

- Refactors Operations to use separate Interfaces to indicate their functional capabilities.
- Refactors Migration/Operation Processors to indicate handling of all operation types with conditional processing based on the functional capabilities of the operations reported by the operations themselves
- Renames the Migration class to TableMigration (for better clarity of purpose)
- Adds a Migration alias class which extends TableMigration for backwards compatibility. Note: This will be deprecated in a future release.
- Updates composer to reflect min PHP standards of 7.4 or above
- Updates/unifies syntax standards for interfaces, parameters, type hinting, etc. to PHP 7.4 standards

Glenn-Ewing reviewed (commented):
Self-review completed.

joshmcrae committed:
Extended new operation interfaces from original

Also removed knowledge of new interfaces from AbstractOperation

joshmcrae committed:
Fixed documentation

joshmcrae commented:
@Glenn-Ewing

Thanks for refactoring this with interface segregation. I've decided to remove the interface-aware methods from `AbstractOperation`, using `instanceof` where needed instead. I've also made the new interfaces extend the base interface since they will always be used in the context of base operations.

I've also gone ahead and changed some document level comments you've included into DocBlocks with `@deprecated` params. This will hint to IDE users that the class is to be replaced.

There were some remnant references to `Migration` and the `MigrationTable` class so I've updated those and also made method return signatures consistent for builder methods. There was a mix of `self`, `$this`, `TableMigration` and `Migration` across DocBlocks and 7.4 type-hints.

Please let me know if you have any concerns after my changes, but I'm ready to get this merged if you're happy with it.

Glenn-Ewing reviewed (commented):
Change review complete. Looks good @joshmcrae !

Glenn-Ewing commented:
@joshmcrae I am ashamed to say that I didn't realize PHP supported interface extension, otherwise I have done what you did, so I am very glad to see it done, thanks! I guess the good news is: I learned something new today, so I can check that off the list; thanks for that! 😄 

I'm totally good with simple instanceof checks, I was making assumptions based on conversation, but am great with simple.

Good with the other DocBlock updates as well, thanks for those, sorry to have missed them.

Im good to merge & release if you are!

joshmcrae commented:
Released as [2.0.0](https://github.com/tithely/exo/releases/tag/2.0.0).
